### DISCUSSION TITLE
Accessing all key bindings via i3-msg?

### DISCUSSION THREAD

**Discussion by RoUS** [i3/i3#4971] [Q&A]
Is there any way to get a dump of _all_ the current `i3` key bindings, both custom and those that are preconfigured but not overridden?  Getting the config file doesn't help because that only lists the custom settings, and in the command language at that.  I'm hoping to build a YAML file of **all** of the bindings, and binding modes, for context-sensitive help.

Thanks!

**Comment by Airblader** [comment]
If by preconfigured you mean interactions like resizing with right click, then no, those are not exposed in any way (and can't even be expressed as a binding).

**Reply by RoUS** [reply]
No, I mean all the default key bindings that would obtain if I didn't include any in my config file.  I want to get the currently __defined__ bindings, which would be those defaults, including any selectively overwritten by any bindings I _do_ make in my config.  The other interactions I'm not interested in (at least not ATM) -- just key bindings.

So, the complete set of bindings `i3` has when it starts up, modified by whatever my config file adds/changes.

Well, I'd also like to know what the assignments are for`Mod1` through `Mod4`, too. :-)

Thanks!

**Reply by Airblader** [reply]
Can you give an example of a binding that is defined by i3 and not through a config file that you are referring to?

The answer is still "no", regardless. There is no mechanism in the IPC to retrieve anything other than the config file.

**Reply by RoUS** [reply]
How about mod-Shift-q and mod-Shift-r? They're defined by default; I don't have to do anything in my config file for those keys to have those bindings active.

Are FREQs and PRs reported on Github? If the base i3 code language is one I know, perhaps I'll put together a patch. What's the proper procedure for this?

Thanks!

**Reply by Airblader** [reply]
> How about mod-Shift-q and mod-Shift-r? They're defined by default; I don't have to do anything in my config file for those keys to have those bindings active.

Those are defined in the default config, they're not hard coded into i3. If you remove them, they will no longer exist. I don't know what your config layout looks like, perhaps your distro ships a default config that includes from another place which is where you define your own overrides. Check `i3 --more-version` to see the "root" file i3 is using.

> Are FREQs and PRs reported on Github?

Yes, you can just open an issue. There'll be a template asking for all required information. But again, what you are referring to is part of the config.
